towhee/dataframe/dataframe.py

```diff
@@ -35,19 +37,171 @@ def __init__(self, name: str, data: List[Tuple[Variable]] = None):
                 `Dataframe`. These tuples can be interpreted as being direct outputs
                 into downstream operators.
         """
-        # self._iter = ScalarIterator(df)
         self._name = name
-        self._data = data
+        self._data = data if data is not None else []
+        self._registered_iter = []
+        self._start_index = 0
+        self._total = 0
+        self._end = False
+
+        self._lock = threading.Lock()
+        self._registered_lock = threading.Lock()
 
     @property
     def name(self):
         return self._name
 
-    def __getitem__(self, val):
-        return self.data[val]
+    def put(self, item: Tuple[Variable]) -> None:
+        assert not self._end, 'DataFrame %s already end, can not put data' % self._name
+        with self._lock:
+            self._data.append(item)
+            self._total += 1
+
+    def get(self, start: int, count: int, force_size: bool = False) -> Tuple[bool, List[Tuple[Variable]]]:
+        """
+        Get [start: start + count) items of dataframe
+        `force_size` is true, return data size must equal `count`. If there is
+        not enough data, return empty list.
+        `force_size` is false, return no more than `count` data
+
+        Returns:
+            (is_end, data)
+        """
+        # TODO (junjie.jiang) call gc function
+        with self._lock:
```

Comment:
we need further optimization for`get` `put` `seal` based on the following logic. (we refactor this part of codes after v0.1)
* put and seal will be called by the same operator
* seal will always be called after put
* iterators never ends until they meet `df._sealed == True`

Reply:
Add TODO for this